packages/docusaurus-theme-common/src/contexts/tabGroupChoice.tsx

```diff
@@ -80,10 +142,56 @@ export function TabGroupChoiceProvider({
   return <Context.Provider value={value}>{children}</Context.Provider>;
 }
 
-export function useTabGroupChoice(): ContextValue {
+type UseTabGroupChoice = {
+  selectedValue: string | null;
+  setTabGroupChoice: (newChoice: string) => void;
+};
+
+export function useTabGroupChoice(
+  groupId: string | undefined,
+  queryString: string | boolean | undefined,
+  defaultValue: string | null,
+  values: readonly {value: string}[],
+): UseTabGroupChoice {
+  const searchKey = getSearchKey(groupId, queryString);
+  const [selectedValue, setSelectedValue] = useState<string | null>(
+    defaultValue,
+  );
   const context = useContext(Context);
   if (context == null) {
     throw new ReactContextError('TabGroupChoiceProvider');
   }
-  return context;
+
+  const setTabGroupChoice = useCallback<UseTabGroupChoice['setTabGroupChoice']>(
+    (newChoice) => {
+      setSelectedValue(newChoice);
+      context.setTabGroupChoice(groupId, queryString, newChoice);
+    },
+    [context, groupId, queryString],
+  );
+
+  // Sync storage, query params, and selected state.
+  useEffect(() => {
+    const queryParamValue =
+      searchKey && context.tabGroupChoicesInQueryParams[searchKey];
+    const storageValue = groupId && context.tabGroupChoicesInStorage[groupId];
+    const valueToSync = queryParamValue ?? storageValue;
+    const isValid =
+      !!valueToSync && values.some(({value}) => value === valueToSync);
+    if (isValid && valueToSync !== selectedValue) {
+      setSelectedValue(valueToSync);
+    }
+  }, [
```

Comment:
**Essence of this PR**: Detect changes in local storage or query params, then update `selectedValue` accordingly.